### PR TITLE
feat: Handle env vars according to environment

### PR DESCRIPTION
**Pull request by hugotiburtino**
That way we also document which environment variables are used in the deployments

- [x] before deploying add `ENVIRONMENT` in the `.env` files of development and staging environment

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -31,3 +31,14 @@ BUCKET_REGION=eu-central-3
...
+# LTI platform: itslearning (staging)
+ITSLEARNING_URL=
+SERLO_EDITOR_CLIENT_ID_ON_ITSLEARNING=
+
+# LTI platform: edu-sharing (RLP) (staging)
+EDUSHARING_RLP_URL=
+SERLO_EDITOR_CLIENT_ID_ON_EDUSHARING_RLP=
+
+# LTI tool: edu-sharing (RLP) (staging)
+EDUSHARING_RLP_CLIENT_ID_ON_SERLO_EDITOR=
```
Comment (hugotiburtino) [inline .env.template]: All these variables aren't secrets. Even the client Id can been seen in the url in the OIDC flow.
We could in principle hard code it. 
It sounds though strange to expose the staging domain of our partners to the rest of the world.

**Review by hejtful — APPROVED** [review]

File: .env.template (hunk 1 above)
Comment (LarsTheGlidingSquirrel) [reply .env.template]: I had the same thought and decided to handle them like (more or less) secrets for now. I would leave them in `.env`.

**Review by LarsTheGlidingSquirrel — COMMENTED** [review]
Thank you for this idea! I see where you are going with this. 
To me, the readability of the env type would improve if there was one per environment (and no shared base type). That comes with some code duplication as a downside but is more readable to me personally and requires less knowledge about `io-ts`. 
Also, I prefer to have the full endpoints in `.env` instead of building them in code. Similar thought here.

**Comment by hejtful** [comment]
From my POV, shared base type for config is what we already use in Serlo, so it's not a new thing introduced here. For example, eslint config in frontend monorepo.

Same for `io-ts`, we use it all around.

**Comment by hugotiburtino** [comment]
> Also, I prefer to have the full endpoints in .env instead of building them in code.

Ok, I've reverted it.

**Comment by hugotiburtino** [comment]
> To me, the readability of the env type would improve if there was one per environment (and no shared base type)

I suppose the main part that is hard to read is this one:

```typescript
const IOEnv = t.union([
  t.intersection([BaseEnv, t.type({ ENVIRONMENT: t.literal('local') })]),
  t.intersection([BaseEnv, t.type({ ENVIRONMENT: t.literal('development') })]),
  t.intersection([
    BaseEnv,
    t.type({ ENVIRONMENT: t.literal('staging') }),
    StagingSpecificEnv,
  ]),
  t.intersection([BaseEnv, t.type({ ENVIRONMENT: t.literal('production') })]),
])
```

But we can't get rid of this union if we want to have env var type checking depending on the environment. Even though, I've refactored it in order to be more readable.

Besides that, the code repetition for each environment would be massive and could be error prone, v.g.

```typescript

const LocalEnv = t.type({
  ENVIRONMENT: t.literal('local'),
  EDITOR_URL: NonEmptyString,
  SERLO_EDITOR_TESTING_SECRET: t.string,
  LTIJS_KEY: NonEmptyString,
  MYSQL_URI: NonEmptyString,
  MONGODB_URI: NonEmptyString,
  ALLOW_SALTIRE: StringOrUndefinedToBoolean,
  ALLOW_EDUSHARING_MOCK: StringOrUndefinedToBoolean,
  S3_ENDPOINT: NonEmptyString,
  BUCKET_NAME: NonEmptyString,
  BUCKET_REGION: NonEmptyString,
  // fallback to '' for now so it does not fail in CI
  BUCKET_ACCESS_KEY_ID: t.string,
  // fallback to '' for now so it does not fail in CI
  BUCKET_SECRET_ACCESS_KEY: t.string,
  MEDIA_BASE_URL: NonEmptyString,
})

const DevelopmentEnv= t.type({
  ENVIRONMENT: t.literal('development'),
  EDITOR_URL: NonEmptyString,
  SERLO_EDITOR_TESTING_SECRET: t.string,
  LTIJS_KEY: NonEmptyString,
  MYSQL_URI: NonEmptyString,
  MONGODB_URI: NonEmptyString,
  ALLOW_SALTIRE: StringOrUndefinedToBoolean,
  ALLOW_EDUSHARING_MOCK: StringOrUndefinedToBoolean,
  S3_ENDPOINT: NonEmptyString,
  BUCKET_NAME: NonEmptyString,
  BUCKET_REGION: NonEmptyString,
  // fallback to '' for now so it does not fail in CI
  BUCKET_ACCESS_KEY_ID: t.string,
  // fallback to '' for now so it does not fail in CI
  BUCKET_SECRET_ACCESS_KEY: t.string,
  MEDIA_BASE_URL: NonEmptyString,
})

const StagingSpecificEnv = t.type({
  ENVIRONMENT: t.literal('staging'),
  EDITOR_URL: NonEmptyString,
  SERLO_EDITOR_TESTING_SECRET: t.string,
  LTIJS_KEY: NonEmptyString,
  MYSQL_URI: NonEmptyString,
  MONGODB_URI: NonEmptyString,
  ALLOW_SALTIRE: StringOrUndefinedToBoolean,
  ALLOW_EDUSHARING_MOCK: StringOrUndefinedToBoolean,
  S3_ENDPOINT: NonEmptyString,
  BUCKET_NAME: NonEmptyString,
  BUCKET_REGION: NonEmptyString,
  // fallback to '' for now so it does not fail in CI
  BUCKET_ACCESS_KEY_ID: t.string,
  // fallback to '' for now so it does not fail in CI
  BUCKET_SECRET_ACCESS_KEY: t.string,
  MEDIA_BASE_URL: NonEmptyString,
  ITSLEARNING_NAME: NonEmptyString,
  ITSLEARNING_URL: NonEmptyString,
  ITSLEARNING_AUTHENTICATION_ENDPOINT: NonEmptyString,
  ITSLEARNING_ACCESS_TOKEN_ENDPOINT: NonEmptyString,
  ITSLEARNING_KEYSET_ENDPOINT: NonEmptyString,
  SERLO_EDITOR_CLIENT_ID_ON_ITSLEARNING: NonEmptyString,
  EDUSHARING_RLP_URL: NonEmptyString,
  EDUSHARING_RLP_NAME: NonEmptyString,
  EDUSHARING_RLP_AUTHENTICATION_ENDPOINT: NonEmptyString,
  EDUSHARING_RLP_ACCESS_TOKEN_ENDPOINT: NonEmptyString,
  EDUSHARING_RLP_KEYSET_ENDPOINT: NonEmptyString,
  SERLO_EDITOR_CLIENT_ID_ON_EDUSHARING_RLP: NonEmptyString,
  EDUSHARING_RLP_LOGIN_ENDPOINT: NonEmptyString,
  EDUSHARING_RLP_LAUNCH_ENDPOINT: NonEmptyString,
  EDUSHARING_RLP_DETAILS_ENDPOINT: NonEmptyString,
  EDUSHARING_RLP_CLIENT_ID_ON_SERLO_EDITOR: NonEmptyString,
})

// same for production...
```

I've refactored it a little bit. Please @LarsTheGlidingSquirrel check if it is easier to read now.

PS: If you think that the hard part is also the ones with custom types (`NonEmptyString` and `StringOrUndefinedToBoolean`) io-ts guy has also the library `io-ts-types` that hides some of these logics. I didn't add it in order to spare us from another dependency, but if you prefer we can add it and simplify the code.

File: .env.template [hunk 2]
```diff
@@ -31,3 +31,25 @@ BUCKET_REGION=eu-central-3
...
+EDUSHARING_RLP_AUTHENTICATION_ENDPOINT="$EDUSHARING_RLP_URL/rest/ltiplatform/v13/auth"
+EDUSHARING_RLP_ACCESS_TOKEN_ENDPOINT="$EDUSHARING_RLP_URL/rest/ltiplatform/v13/token"
+EDUSHARING_RLP_KEYSET_ENDPOINT="$EDUSHARING_RLP_URL/rest/lti/v13/jwks"
+SERLO_EDITOR_CLIENT_ID_ON_EDUSHARING_RLP=
+
+# LTI tool: edu-sharing (RLP) (staging)
+EDUSHARING_RLP_LOGIN_ENDPOINT="$EDUSHARING_RLP_URL/rest/lti/v13/oidc/login_initiations"
+EDUSHARING_RLP_LAUNCH_ENDPOINT="$EDUSHARING_RLP_URL/rest/lti/v13/lti13"
+EDUSHARING_RLP_DETAILS_ENDPOINT="$EDUSHARING_RLP_URL/rest/lti/v13/details"
+EDUSHARING_RLP_CLIENT_ID_ON_SERLO_EDITOR=
```
Comment (LarsTheGlidingSquirrel) [inline .env.template:55]: What is the reasoning behind adding them to the template? 
In environment `local` we cannot test with them sadly. 
I assumed `.env.template` is just for `local`.

File: src/backend/register-lti-platforms.ts [hunk 3]
```diff
@@ -23,63 +23,42 @@ export async function registerLtiPlatforms() {
...
-    process.env.ITSLEARNING_KEYSET_ENDPOINT
-  ) {
-    const platform = await ltijs.registerPlatform({
-      url: process.env.ITSLEARNING_URL, // LTI iss
-      name: process.env.ITSLEARNING_NAME,
-      clientId: process.env.SERLO_EDITOR_CLIENT_ID_ON_ITSLEARNING,
-      authenticationEndpoint: process.env.ITSLEARNING_AUTHENTICATION_ENDPOINT,
-      accesstokenEndpoint: process.env.ITSLEARNING_ACCESS_TOKEN_ENDPOINT,
+  if (config.ENVIRONMENT === 'staging') {
+    const itsLearningPlatform = await ltijs.registerPlatform({
```
Comment (LarsTheGlidingSquirrel) [inline src/backend/register-lti-platforms.ts]: Comment can move inside the if block.

**Review by LarsTheGlidingSquirrel — APPROVED** [review]
The env types look way clearer now! Thank you for improving it! 
Apart from a few small comments (no blocker) everything looks fine to me. :relaxed:

File: .env.template (hunk 2 above)
Comment (hugotiburtino) [reply .env.template:55]: To have all these key-values there is mainly for documentation.

In a small extent it makes the life easier for a developer that will deploy the app in a new environment or have to test locally something related to a deployment (in this case, staging).

`.env.template` doesn't have to be just for local environment.

**Pull request merged**